### PR TITLE
Fix BERT FSDP test

### PR DESCRIPTION
**Pull request by regisss**
# What does this PR do?

This PR fixes a bug where the FSDP wrapper of the model is lost at some point in Accelerate:
- Distributed wrappers (i.e. DDP, FSDP, DeepSpeed) are removed [here](https://github.com/huggingface/accelerate/blob/v0.33.0/src/accelerate/utils/other.py#L90),
- At some point after that operation is performed, the FSDP wrapper is lost, which makes the BERT FSDP test fail.

This PR fixes that issue by keeping a buffer variable to not lose track of the distributed model and making sure that it is set it back at the end.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

**Comment by regisss** [comment]
Test fails because I get a throughput of 3234.769 and it should be larger than 3516.322 though

**Comment by vidyasiv** [comment]
Thanks @regisss , I see the test passing and performance is as follows:
```
***** train metrics *****
  epoch                       =        2.0
  max_memory_allocated (GB)   =       6.85
  memory_allocated (GB)       =        0.4
  total_flos                  =  4047423GF
  total_memory_available (GB) =      94.62
  train_loss                  =     1.3817
  train_runtime               = 0:01:02.51
  train_samples               =      88524
  train_samples_per_second    =   3602.047
  train_steps_per_second      =     18.803
***** eval metrics *****
  epoch                       =        2.0
  eval_exact_match            =    76.8874
  eval_f1                     =    85.3401
  eval_runtime                = 0:00:13.00
  eval_samples                =      10784
  eval_samples_per_second     =    829.419
  eval_steps_per_second       =     12.998
  max_memory_allocated (GB)   =       6.85
  memory_allocated (GB)       =       0.76
  total_memory_available (GB) =      94.62
```
I have closed my other PR: https://github.com/huggingface/optimum-habana/pull/1274

**Review by yafshar — APPROVED** [review]
LGTM!

thanks @regisss, very nice catch! I missed that the FSDP wrapper is lost!

File: optimum/habana/accelerate/utils/other.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+        from deepspeed import DeepSpeedEngine
+
+        options += (DeepSpeedEngine,)
+
+    if is_torch_version(">=", FSDP_PYTORCH_VERSION) and is_torch_distributed_available():
+        from torch.distributed.fsdp.fully_sharded_data_parallel import FullyShardedDataParallel as FSDP
+
+        options += (FSDP,)
+
+    distributed_model = model
```
Comment (vivekgoe) [inline optimum/habana/accelerate/utils/other.py:38]: Looks good solution to me for now. But we will have problem in future if some use-case needs functionality from L42 - L69, with this change we are effectively skipping those lines. Maybe good to add a comment warning about this.

**Review by vivekgoe — COMMENTED** [review]
@regisss reviewed change, please check comments. Thanks.

**Comment by yafshar** [comment]
@regisss do you know or can you comment why the original solution is not an issue on non-Gaudi devices?

**Comment by regisss** [comment]
> @regisss do you know or can you comment why the original solution is not an issue on non-Gaudi devices?

It's still unclear to me why it works on GPU, I'll need to take a closer look at it.

**Review by vivekgoe — COMMENTED** [review]
LGTM.

**Pull request merged**